src/api/java/appeng/api/networking/crafting/ICraftingPatternDetails.java

```diff
@@ -60,24 +62,61 @@
     boolean isCraftable();
 
     /**
-     * @return a list of the inputs, will include nulls.
+     * Equal itemstacks will be aggregated into one, respectively 3*64 will be
+     * returned as one stack of 192, up to 576 of a single type.
+     * 
+     * This should be the preferred way to deal with the list of inputs.
+     * 
+     * There is no guarantee about maintaining the placement order of the input or
+     * output slots.
+     * 
+     * @return an immutable list of inputs without nulls
      */
-    IAEItemStack[] getInputs();
+    List<IAEItemStack> getInputs();
 
     /**
-     * @return a list of the inputs, will be clean
+     * * Equal itemstacks will be aggregated into one, respectively 2*32 will be
+     * returned as one stack of 64, up to 192 of a single type.
+     * 
+     * This should be the preferred way to deal with the list of outputs.
+     * 
+     * There is no guarantee about maintaining the placement order of the input or
+     * output slots.
+     * 
+     * @return an immutable list of outputs without nulls
      */
-    IAEItemStack[] getCondensedInputs();
+    List<IAEItemStack> getOutputs();
 
     /**
-     * @return a list of the outputs, will be clean
+     * A sparse list representing the placement order of a crafting grid.
+     * 
+     * Only use when absolutely necessary, always prefer
+     * {@link ICraftingPatternDetails#getInputs()}
+     * 
+     * This will contain exactly 9 entries.
+     * 
+     * This can return a copy from the internal structure, so there are no
+     * guarantees about modifications.
+     * 
+     * @return a list of the inputs, will include nulls.
      */
-    IAEItemStack[] getCondensedOutputs();
+    IAEItemStack[] getSparseInputs();
 
     /**
+     * A sparse list representing the placement order of the respective output
+     * slots.
+     * 
+     * Only use when absolutely necessary, always prefer
+     * {@link ICraftingPatternDetails#getOutputs()}
+     * 
+     * This will either contain 1 entry for crafting patterns or 3 for processing.
+     * 
+     * This can return a copy from the internal structure, so there are no
+     * guarantees about modifications.
```

Comment:
```suggestion
     * A sparse list representing the placement order of the respective output
     * slots.
     * <p>
     * Only use when absolutely necessary, always prefer
     * {@link ICraftingPatternDetails#getOutputs()}
     * <p>
     * This will either contain 1 entry for crafting patterns or 3 for processing.
     * <p>
     * This can return a copy from the internal structure, so there are no
     * guarantees about modifications.
```